Sharing page, explaining the `shared` keyword and pubsub examples (#79)
dfinity/docs: modules/language-guide/pages/sharing.adoc

= Sharing data and behavior

Recall that in {proglang}, mutable state is always private to an actor.

However, two actors can share message data, and those messages can

> refer to actors, including themselves and one another.  Additionally,

Reviewer: ```suggestion
refer to actors, including the communicating and other actors.  Additionally,
```

messages may refer to individual functions, if those functions are `shared`.

Through these mechanisms, two actors can coordinate their behavior
through asynchronous message passing.

== Publisher-subscriber pattern with actors

> We show how actors share their functions by focusing on variations of the link:https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern["publisher-subscriber" pattern],

Reviewer: ```suggestion
We demonstrate how actors may share their functions by focusing on variations of the link:https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern["publish-subscribe" pattern] (of contracted to  "pubsub"),
```

Reviewer 2: "actors can share" instead of may
(referred to as "pubsub")

wherein a "publisher" actor records a list of "subscriber" actors to
notify when something notable occurs in the publisher's state (e.g., a
new "publication" is available).

Below, we use two actors in {proglang} to build variations of the

> publisher-subscriber relationship between the actors.

Reviewer: ```suggestion
publish-subscrib relationship between the actors.
```

Reviewer 2: replace "e.g." with "for example,"

Reviewer 2: publish-subscribe

For a complete working project,
see the link:https://github.com/dfinity/examples/tree/master/motoko/pubsub[pubsub example].

> Below, we introduce the ideas gradually, showing the most salient pieces in code below.

Reviewer: ```suggestion
Below, we introduce the ideas gradually, showing the most salient pieces in code.
```

> Below, we introduce the ideas gradually, showing the most salient pieces in code below.

Reviewer: ```suggestion
We introduce the ideas gradually, showing the most salient pieces in code below.
```
You're bellowing ;->

Author: That's a thing?

See the link:https://github.com/dfinity/examples[`examples` repository] for working code.

=== A basic version

> Consider the following `Subscriber` actor type, which gives a possible
> interface for the subscriber actor and publisher actor to expose and
> to call, respectively:

Reviewer: ```suggestion
Consider the following `Subscriber` actor type. This presents an agreed
interface that a subscribing actor can expose for a publishing actor to call.
```

Reviewer 2: agreed-upon

[source,motoko]
----
public type Subscriber = actor {
  notify : () -> ()
};
----

1. The `Publisher` uses this type to define a data structure to store its
subscribers as data.
2. Each `Subscriber` actor exposes an update function

> `notify`, as in the `Subscriber` actor type signature above.  Recall that

Reviewer: ```suggestion
`notify` implementing `Subscriber` actor type, or interface, above.  Recall that
```

the subscribers receive the `notify` message, as in the `Subscriber`
type given above.

=== Subscriber side

In the simplest case, the subscriber side has three methods:
1. initially, subscribe to notifications from the publisher (`init`),
2. be notified as one of the subscribed actors (`notify`, as in the `Subscriber` type given above), and
1. permit queries to the accumulated state, in this case just a counter of notifications (`get`).

The code below implements each of these:

[source,motoko]
----
import Publisher "canister:pub";

actor Subscriber {
    var count: Nat = 0;
    public func init() {
        Publisher.subscribe(Subscriber);
    };
    public func notify() {
        count += 1;
    };
    public func get() : async Nat {
        count
    };
}
----

The actor assumes, but does not enforce, that its `init` function is
only ever called once.  In it, the `Subscriber` actor passes a
reference to itself, of type `actor { notify : () -> () };` (locally
called `Subscriber` above).

If called more than once, the actor will subscribe itself multiple
times, and will receive multiple (duplicate) notifications from the
publisher.  This fragility is the consequence of the basic

> publiisher-subscriber design we show above.  With more care, a more

Reviewer: ```suggestion
publisher-subscriber design we show above.  With more care, a more
```

advanced publisher actor could check for duplicate subscriber actors
and ignore them, for instance.

== Sharing functions among actors

In {proglang}, a `shared` actor function can be sent in a message to
another actor, and then later called by that actor, or another one.

The code shown above has been simplified for illustrative purposes.
The full version offers additional features to the
publisher-subscriber relationship, and uses shared functions to make
this relationship more flexible.

> For instance, the notification function is _always_ designated as
> `notify`.  A more flexible design would only fix the type of `notify`,
> and permit the subscriber to choose any of its `shared` functions,
> specified in a `subscribe` message in place of (just) the actor that
> is subscribing.

Reviewer: ```suggestion
For instance, the notification function is an actor member named 
`notify`.  A more flexible design would only fix the signature of some notification function,
and permit the subscriber to choose any of its `shared` functions that match the signature, regardless of name,
specified in a `subscribe` message as an individual function rather than some containing actor.
This would allow an actor to register two different notification functions with the same or even separate publishers, for example.
```

See the link:https://github.com/dfinity/examples/tree/master/motoko/pubsub[the full example]
for details.

In particular, suppose that the subscriber wants to
avoid being locked into a certain naming scheme for its interface.
What really matters is that the publisher can call _some_ function
that the subscriber chooses.

=== The keyword `shared`

To permit this flexibility, an actor needs to share a single
_function_ that permits remote invocation from another actor, not
merely a reference to itself.

The ability to share a function requires that it be pre-designated as
`shared`, and the type system enforces that these functions follow
certain rules around the types of data that these functions accept,

> return and over which their closures close.

Reviewer: ```suggestion
return and access through their code.
```

> We omit this keyword for actor methods since implicitly, _all public
> functions of every actor are `shared`_, whether marked explicitly
> or not.

Reviewer 3: ```suggestion
As a shorthand, Motoko allows to omit this keyword for _public_ actor methods since implicitly, _all public
functions of every actor are `shared`_.
```

> More generally, a `shared` function is one that is _either_ part of
> the public interface of an actor, _or_ it is not mentioned in the
> public interface, but it does not close over the actor's mutable
> state, and it adheres to the same argument and return-type typing
> restrictions as a public actor function (e.g., no mutable data in
> the arguments or results).

Reviewer: ```suggestion
More generally, a `shared` function is one that is _either_ locally defined as part of
the public interface of the current actor, _or_  was received as a parameter to a method.
In either case, it adheres to the same argument and return-type typing
restrictions as a locally defined actor function (e.g., no mutable data in
the arguments or results).
```

> More generally, a `shared` function is one that is _either_ part of
> the public interface of an actor, _or_ it is not mentioned in the
> public interface, but it does not close over the actor's mutable
> state, and it adheres to the same argument and return-type typing
> restrictions as a public actor function (e.g., no mutable data in
> the arguments or results).

Reviewer 3: ```suggestion
More generally, a `shared` function is one that is _either_ part of
the public interface of an actor, _or_ a private function that is explicitly declared `shared`. In either case, it adheres to the same argument and return-type typing
restrictions as a public actor function (e.g., no mutable data in
the arguments or results).
```
@crusso, do we also support local shared functions on the platform? In the interpreter, the following is fine, but I suppose we cannot do that yet:
```
> actor B { public func f() : async(shared ()->()) { shared func () {}}};
let B : actor {f : shared () -> async shared () -> ()} = {f = func}
> async (await B.f())();
async () : async<$top-level> ()
```

Using the `shared` type, we can extend the example above to be more
flexible:

[source,motoko]
----

> type SubscribeMessage = { callback: shared () -> (); };

Reviewer: Why do you wrap it in record type? Why not just define:
```
type SubscribeMessage = shared () -> ();
```

Reviewer: I think everything would still work and be simpler.

Author: Good point.  I adapted it from Stanley's example, which had an additional field in the record that I omitted here.
